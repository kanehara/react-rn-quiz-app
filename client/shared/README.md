# The shared module for the trivia clients
A module housing common JS code between the web and React Native clients.

## Contains
* Redux actions, reducers, getters
* Apollo client used in redux layer to fetch data
* styled components theme