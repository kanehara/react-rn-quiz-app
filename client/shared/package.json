{
  "name": "shared",
  "version": "0.0.1",
  "scripts": {
    "test": "jest"
  },
  "devDependencies": {
    "@babel/core": "^7.2.2",
    "@babel/preset-env": "^7.3.1",
    "babel-jest": "^24.0.0",
    "jest": "^24.0.0",
    "unfetch": "^4.0.1"
  }
}
