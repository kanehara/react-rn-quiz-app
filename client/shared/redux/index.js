import { createStore, applyMiddleware, compose } from 'redux'
import thunk from 'redux-thunk'
import { quizReducer } from './quiz'
import { persistStore, persistReducer } from 'redux-persist'
import storage from 'redux-persist/lib/storage'

const composeEnhancers =
  typeof window === 'object' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ 
    ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({}) 
    : compose

const persistedReducer = persistReducer({
  key: 'root',
  storage
}, quizReducer)

const store = createStore(
  persistedReducer,
  composeEnhancers(
    applyMiddleware(thunk)
  )
)

const persistor = persistStore(store)

export {
  store,
  persistor
}
