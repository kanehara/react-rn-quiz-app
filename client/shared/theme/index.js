import theme from './theme'
import { ThemeProvider as T } from 'styled-components'
import React from 'react'

const ThemeProvider = ({children}) => <T theme={theme}>{children}</T>

export {
  ThemeProvider
}
