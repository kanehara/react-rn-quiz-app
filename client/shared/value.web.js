export default 'value in web2'
