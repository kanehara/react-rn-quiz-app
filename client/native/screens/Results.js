import React from 'react'
import {Button, ScrollView} from 'react-native'
import Text from 'native/components/Text'
import ScreenView from './ScreenView'
import { connect } from 'react-redux'
import { getters, actions } from 'shared/redux/quiz'
import styled from 'styled-components'
import {H1, H2} from 'native/components/Header'

const Question = ({question, answer}) => (
  <React.Fragment>
    <Text>
      {question.correct_answer === answer ? 'O ' : 'X '} 
      {question.question} 
      - {question.correct_answer}
    </Text>
  </React.Fragment>
)

class Results extends React.Component {
  componentDidMount () {
    if (this.props.inProgress) {
      this.props.navigation.navigate('quiz')
    }
  }

  reset = () => {
    this.props.reset()
    this.props.navigation.navigate('home')
  }

  render() {
    return (
      <ScreenView>
        <ScrollView>
          <H1>Results</H1>
          <H2>{this.props.score} / {this.props.questions.length}</H2>
          {this.props.questions.map((q, i) => 
            <Question key={i} question={q} answer={this.props.answers[i]} />
          )}
        </ScrollView>
        <Button onPress={this.reset} title='Restart' />
      </ScreenView>
    );
  }
}

export default connect(
  state => ({
    score: getters.score(state),
    finished: getters.finished(state),
    inProgress: getters.inProgress(state),
    answers: state.answers,
    questions: state.questions
  }),
  dispatch => ({
    reset: () => dispatch(actions.reset())
  })
)(Results)
