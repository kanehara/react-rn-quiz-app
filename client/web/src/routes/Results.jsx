import React from 'react'
import { connect } from 'react-redux'
import { Redirect } from 'react-router-dom'
import styled from 'styled-components'
import Button from '../components/Button'
import {actions, getters} from 'shared/redux/quiz'
import {Flex, Ray} from 'rayout'

const QuestionsContainer = styled.div`
  margin: auto;
  margin-bottom: 40px;
`

const QuestionHeader = styled.h3`
  color: ${({ correct, theme }) => correct 
    ? theme.correctColor
    : theme.incorrectColor
  };
  margin-bottom: 5px;
  margin-left: 20px;
`

class Question extends React.Component {
  render () {
    const {question, answer} = this.props
    const correct = answer === question.correct_answer
    return (
      <Flex flexDirection='column' textAlign='left'>
        <Flex>
          <h1>{correct ? '✅' : '❌'}</h1>
          <QuestionHeader correct={correct}>
            {question.question} - {question.correct_answer}
          </QuestionHeader>
        </Flex>
      </Flex>
    )
  }
}

export class Results extends React.Component {
  static defaultProps = {
    questions: []
  }

  restart = () => {
    this.props.reset()
    this.props.history.replace('/')
  }

  render () {
    if (this.props.finished) {
      return (
        <div data-testid='homeRedirect'>
          <Redirect to='/' />
        </div>
      )
    } else if (this.props.inProgress) {
      return (
        <div data-testid='quizRedirect'>
          <Redirect to='/quiz' />
        </div>
      )
    }

    return (
      <Ray maxWidth='50%' margin='auto'>
        <h2>You scored</h2>
        <h2>{this.props.score} / {this.props.questions.length}</h2>
        <QuestionsContainer>
          {this.props.questions.map((q, i) => (
            <Question answer={this.props.answers[i]} question={q} key={i} />
          ))}
        </QuestionsContainer>
        <Button onClick={this.restart}>Restart</Button>
      </Ray>
    )
  }
}

export default connect(
  state => ({
    score: getters.score(state),
    inProgress: getters.inProgress(state),
    finished: getters.finished(state),
    answers: state.answers,
    questions: state.questions
  }),
  dispatch => ({
    reset: () => dispatch(actions.reset())
  })
)(Results)
