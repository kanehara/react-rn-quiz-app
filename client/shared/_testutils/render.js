import React from 'react'
import { ThemeProvider } from 'styled-components'
import theme from 'shared/theme'
import { Provider } from 'react-redux' 
import { store } from 'shared/redux'
import renderer from 'react-test-renderer'

export const render = jsx => renderer.create(
  <Provider store={store}>
    <ThemeProvider theme={theme}>
      {jsx}
    </ThemeProvider>
  </Provider>
)

