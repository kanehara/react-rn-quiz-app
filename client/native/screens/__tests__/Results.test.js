import {Results} from '../Results'
import React from 'react'
import { render } from 'shared/_testutils'
import {TouchableHighlight} from 'react-native'

describe('Results', () => {
  const getMockNavigation = () => ({
    navigate: jest.fn()
  })

  it('renders', () => {
    expect(render(
      <Results 
        questions={[
          {correct_answer: 'true', question: 'question0'},
          {correct_answer: 'true', question: 'question1'},
          {correct_answer: 'false', question: 'question2'}
        ]} 
        answers={['true', 'false', 'true']} 
      />
    )).toMatchSnapshot()
  })
  
  it('navigates to quiz if in progress', () => {
    const mockNav = getMockNavigation()
    render(<Results navigation={mockNav} inProgress />)
    expect(mockNav.navigate).toHaveBeenCalledWith('quiz')
  })
  
  it('resets quiz on button press', async () => {
    const mockNav = getMockNavigation()
    const resetMock = jest.fn()
    const instance = render(<Results navigation={mockNav} reset={resetMock} />)
    const button = instance.root.findByType(TouchableHighlight)
    button.props.onPress()
    expect(resetMock).toHaveBeenCalled()
    expect(mockNav.navigate).toHaveBeenCalledWith('home')
  })
})
