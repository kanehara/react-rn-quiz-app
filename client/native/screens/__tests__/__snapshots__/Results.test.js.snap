// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Results renders 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#f5faf4",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "center",
        "paddingBottom": 40,
        "paddingLeft": 20,
        "paddingRight": 20,
        "paddingTop": 40,
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#292F36",
          "fontSize": 28,
          "fontWeight": "bold",
          "marginBottom": 16,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 4,
          "textAlign": "center",
        },
      ]
    }
  >
    You scored
  </Text>
  <Text
    style={
      Array [
        Object {
          "color": "#292F36",
          "fontSize": 22,
          "fontWeight": "bold",
          "marginBottom": 12,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 4,
          "textAlign": "center",
        },
      ]
    }
  >
     / 
    3
  </Text>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#ffffff",
          "borderColor": "#c5e0de",
          "borderRadius": 10,
          "borderStyle": "solid",
          "borderWidth": 1,
          "justifyContent": "center",
          "marginBottom": 10,
          "marginLeft": 10,
          "marginRight": 10,
          "marginTop": 10,
          "paddingBottom": 5,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 5,
          "width": "100%",
        },
      ]
    }
  >
    <View
      style={null}
    >
      <Text
        style={
          Array [
            Object {
              "color": "#292F36",
              "fontSize": 16,
              "fontWeight": "700",
              "marginBottom": 4,
              "marginLeft": 0,
              "marginRight": 0,
              "marginTop": 4,
            },
          ]
        }
      >
        Restart
      </Text>
    </View>
  </View>
</View>
`;
